File: flutter_app/lib/xc_economy.dart
import 'package:shared_preferences/shared_preferences.dart';
import 'dart:convert';

class XCTransaction {
  final String id;
  final String type; // 'earn' | 'spend' | 'bonus' | 'achievement'
  final int amount;
  final String description;
  final int timestamp;
  final String source;

  XCTransaction({
    required this.id,
    required this.type,
    required this.amount,
    required this.description,
    required this.timestamp,
    required this.source,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'type': type,
      'amount': amount,
      'description': description,
      'timestamp': timestamp,
      'source': source,
    };
  }

  factory XCTransaction.fromJson(Map<String, dynamic> json) {
    return XCTransaction(
      id: json['id'],
      type: json['type'],
      amount: json['amount'],
      description: json['description'],
      timestamp: json['timestamp'],
      source: json['source'],
    );
  }
}

class XCAchievement {
  final String id;
  final String name;
  final String description;
  final int reward;
  final bool completed;
  final int progress;
  final int maxProgress;
  final String icon;

  XCAchievement({
    required this.id,
    required this.name,
    required this.description,
    required this.reward,
    required this.completed,
    required this.progress,
    required this.maxProgress,
    required this.icon,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'reward': reward,
      'completed': completed,
      'progress': progress,
      'maxProgress': maxProgress,
      'icon': icon,
    };
  }

  factory XCAchievement.fromJson(Map<String, dynamic> json) {
    return XCAchievement(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      reward: json['reward'],
      completed: json['completed'] ?? false,
      progress: json['progress'] ?? 0,
      maxProgress: json['maxProgress'] ?? 1,
      icon: json['icon'] ?? '🏆',
    );
  }
}

class XCStreak {
  final int currentStreak;
  final String lastActiveDate;
  final int totalDays;
  final int nextBonus;

  XCStreak({
    required this.currentStreak,
    required this.lastActiveDate,
    required this.totalDays,
    required this.nextBonus,
  });

  Map<String, dynamic> toJson() {
    return {
      'currentStreak': currentStreak,
      'lastActiveDate': lastActiveDate,
      'totalDays': totalDays,
      'nextBonus': nextBonus,
    };
  }

  factory XCStreak.fromJson(Map<String, dynamic> json) {
    return XCStreak(
      currentStreak: json['currentStreak'] ?? 0,
      lastActiveDate: json['lastActiveDate'] ?? '',
      totalDays: json['totalDays'] ?? 0,
      nextBonus: json['nextBonus'] ?? 10,
    );
  }
}

class XCEconomy {
  static XCEconomy? _instance;
  static XCEconomy get instance => _instance ??= XCEconomy._();

  XCEconomy._();

  int _balance = 1240;
  List<XCTransaction> _transactions = [];
  List<XCAchievement> _achievements = [];
  XCStreak _streak = XCStreak(
    currentStreak: 0,
    lastActiveDate: '',
    totalDays: 0,
    nextBonus: 10,
  );

  int get balance => _balance;
  List<XCTransaction> get transactions => List.unmodifiable(_transactions);
  List<XCAchievement> get achievements => List.unmodifiable(_achievements);
  XCStreak get streak => _streak;

  Future<void> _loadState() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final saved = prefs.getString('xc_economy');
      if (saved != null) {
        final data = json.decode(saved);
        _balance = data['balance'] ?? 1240;
        _transactions = (data['transactions'] as List?)
                ?.map((t) => XCTransaction.fromJson(t))
                .toList() ??
            [];
        _streak = XCStreak.fromJson(data['streak'] ?? {});
      }
    } catch (error) {
      print('Failed to load XC economy state: $error');
    }
  }

  Future<void> _saveState() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString('xc_economy', json.encode({
        'balance': _balance,
        'transactions': _transactions.map((t) => t.toJson()).toList(),
        'streak': _streak.toJson(),
      }));
    } catch (error) {
      print('Failed to save XC economy state: $error');
    }
  }

  void _initializeAchievements() {
    _achievements = [
      XCAchievement(
        id: 'first_post',
        name: 'First Steps',
        description: 'Post your first buzz message',
        reward: 25,
        completed: false,
        progress: 0,
        maxProgress: 1,
        icon: '📝',
      ),
      XCAchievement(
        id: 'first_chat',
        name: 'Social Butterfly',
        description: 'Send your first chat message',
        reward: 10,
        completed: false,
        progress: 0,
        maxProgress: 1,
        icon: '💬',
      ),
      XCAchievement(
        id: 'daily_login',
        name: 'Regular',
        description: 'Log in 7 days in a row',
        reward: 100,
        completed: false,
        progress: 0,
        maxProgress: 7,
        icon: '📅',
      ),
      XCAchievement(
        id: 'mesh_master',
        name: 'Mesh Master',
        description: 'Connect to 10 different nodes',
        reward: 500,
        completed: false,
        progress: 0,
        maxProgress: 10,
        icon: '🔗',
      ),
    ];
  }

  Future<void> initialize() async {
    await _loadState();
    if (_achievements.isEmpty) {
      _initializeAchievements();
    }
    _checkDailyStreak();
  }

  void addXC(int amount, String description, String source) {
    _balance += amount;
    _transactions.add(XCTransaction(
      id: 'tx_${DateTime.now().millisecondsSinceEpoch}',
      type: amount > 0 ? 'earn' : 'spend',
      amount: amount,
      description: description,
      timestamp: DateTime.now().millisecondsSinceEpoch,
      source: source,
    ));
    _saveState();
  }

  bool spendXC(int amount, String description, String source) {
    if (_balance >= amount) {
      _balance -= amount;
      _transactions.add(XCTransaction(
        id: 'tx_${DateTime.now().millisecondsSinceEpoch}',
        type: 'spend',
        amount: -amount,
        description: description,
        timestamp: DateTime.now().millisecondsSinceEpoch,
        source: source,
      ));
      _saveState();
      return true;
    }
    return false;
  }

  void awardDailyLogin() {
    final today = DateTime.now();
    final todayStr = '${today.year}-${today.month}-${today.day}';
    
    if (_streak.lastActiveDate != todayStr) {
      if (_streak.lastActiveDate == _getYesterday(today)) {
        _streak = XCStreak(
          currentStreak: _streak.currentStreak + 1,
          lastActiveDate: todayStr,
          totalDays: _streak.totalDays + 1,
          nextBonus: _streak.nextBonus + 5,
        );
      } else {
        _streak = XCStreak(
          currentStreak: 1,
          lastActiveDate: todayStr,
          totalDays: _streak.totalDays + 1,
          nextBonus: 10,
        );
      }
      
      // Award streak bonus
      final bonus = _streak.currentStreak * 5;
      addXC(bonus, 'Daily login bonus (streak: ${_streak.currentStreak})', 'daily_bonus');
      _saveState();
    }
  }

  void _checkDailyStreak() {
    final today = DateTime.now();
    final todayStr = '${today.year}-${today.month}-${today.day}';
    
    if (_streak.lastActiveDate != todayStr) {
      if (_streak.lastActiveDate != _getYesterday(today)) {
        // Streak broken
        _streak = XCStreak(
          currentStreak: 0,
          lastActiveDate: _streak.lastActiveDate,
          totalDays: _streak.totalDays,
          nextBonus: 10,
        );
      }
    }
  }

  String _getYesterday(DateTime today) {
    final yesterday = today.subtract(const Duration(days: 1));
    return '${yesterday.year}-${yesterday.month}-${yesterday.day}';
  }

  void updateAchievement(String achievementId, int progress) {
    final achievement = _achievements.firstWhere(
      (a) => a.id == achievementId,
      orElse: () => XCAchievement(
        id: achievementId,
        name: 'Unknown',
        description: 'Unknown achievement',
        reward: 0,
        completed: false,
        progress: 0,
        maxProgress: 1,
        icon: '🏆',
      ),
    );

    if (!achievement.completed) {
      final updatedAchievement = XCAchievement(
        id: achievement.id,
        name: achievement.name,
        description: achievement.description,
        reward: achievement.reward,
        completed: progress >= achievement.maxProgress,
        progress: progress,
        maxProgress: achievement.maxProgress,
        icon: achievement.icon,
      );

      final index = _achievements.indexWhere((a) => a.id == achievementId);
      _achievements[index] = updatedAchievement;

      if (updatedAchievement.completed && !achievement.completed) {
        addXC(achievement.reward, 'Achievement unlocked: ${achievement.name}', 'achievement');
      }
      
      _saveState();
    }
  }

  void completeAchievement(String achievementId) {
    final achievement = _achievements.firstWhere(
      (a) => a.id == achievementId,
      orElse: () => XCAchievement(
        id: achievementId,
        name: 'Unknown',
        description: 'Unknown achievement',
        reward: 0,
        completed: false,
        progress: 0,
        maxProgress: 1,
        icon: '🏆',
      ),
    );

    if (!achievement.completed) {
      updateAchievement(achievementId, achievement.maxProgress);
    }
  }
}

// Global instance for easy access
final xcEconomy = XCEconomy.instance;

File: services/xcEconomy.ts
// XC Economy Service - Virtual Currency System
export interface XCTransaction {
  id: string;
  type: 'earn' | 'spend' | 'bonus' | 'achievement';
  amount: number;
  description: string;
  timestamp: number;
  source: string;
}

export interface XCAchievement {
  id: string;
  name: string;
  description: string;
  reward: number;
  completed: boolean;
  progress: number;
  maxProgress: number;
  icon: string;
}

export interface XCStreak {
  currentStreak: number;
  lastActiveDate: string;
  totalDays: number;
  nextBonus: number;
}

class XCEconomyService {
  private balance: number = 0;
  private transactions: XCTransaction[] = [];
  private achievements: XCAchievement[] = [];
  private streak: XCStreak = {
    currentStreak: 0,
    lastActiveDate: '',
    totalDays: 0,
    nextBonus: 10
  };
  private listeners: { [key: string]: ((data: any) => void)[] } = {};

  constructor() {
    this.loadState();
    this.initializeAchievements();
    this.checkDailyStreak();
  }

  private loadState() {
    try {
      const saved = localStorage.getItem('xc_economy');
      if (saved) {
        const data = JSON.parse(saved);
        this.balance = data.balance || 0;
        this.transactions = data.transactions || [];
        this.streak = data.streak || this.streak;
      }
    } catch (error) {
      console.error('Failed to load XC economy state:', error);
    }
  }

  private saveState() {
    localStorage.setItem('xc_economy', JSON.stringify({
      balance: this.balance,
      transactions: this.transactions,
      streak: this.streak
    }));
  }

  private initializeAchievements() {
    this.achievements = [
      {
        id: 'first_post',
        name: 'First Steps',
        description: 'Post your first buzz message',
        reward: 25,
        completed: false,
        progress: 0,
        maxProgress: 1,
        icon: '📝'
      },
      {
        id: 'buzz_master',
        name: 'Buzz Master',
        description: 'Post 10 buzz messages',
        reward: 50,
        completed: false,
        progress: 0,
        maxProgress: 10,
        icon: '🎯'
      },
      {
        id: 'social_butterfly',
        name: 'Social Butterfly',
        description: 'Interact 100 times',
        reward: 100,
        completed: false,
        progress: 0,
        maxProgress: 100,
        icon: '🦋'
      },
      {
        id: 'mesh_pioneer',
        name: 'Mesh Pioneer',
        description: 'Connect to 10 mesh nodes',
        reward: 75,
        completed: false,
        progress: 0,
        maxProgress: 10,
        icon: '🌐'
      },
      {
        id: 'channel_hopper',
        name: 'Channel Hopper',
        description: 'Join 5 geohash channels',
        reward: 30,
        completed: false,
        progress: 0,
        maxProgress: 5,
        icon: '📍'
      },
      {
        id: 'profile_complete',
        name: 'Profile Complete',
        description: 'Complete your profile setup',
        reward: 20,
        completed: false,
        progress: 0,
        maxProgress: 4,
        icon: '👤'
      },
      {
        id: 'daily_warrior',
        name: 'Daily Warrior',
        description: 'Login for 7 consecutive days',
        reward: 100,
        completed: false,
        progress: 0,
        maxProgress: 7,
        icon: '🗓️'
      },
      {
        id: 'xc_collector',
        name: 'XC Collector',
        description: 'Earn 500 XC total',
        reward: 200,
        completed: false,
        progress: 0,
        maxProgress: 500,
        icon: '💰'
      }
    ];
  }

  private checkDailyStreak() {
    const today = new Date().toDateString();
    const lastActive = this.streak.lastActiveDate;
    
    if (lastActive !== today) {
      const yesterday = new Date(Date.now() - 86400000).toDateString();
      
      if (lastActive === yesterday) {
        // Continue streak
        this.streak.currentStreak++;
        this.streak.totalDays++;
        this.streak.nextBonus = Math.min(10 + this.streak.currentStreak * 5, 50);
      } else {
        // Reset streak
        this.streak.currentStreak = 1;
        this.streak.nextBonus = 10;
      }
      
      this.streak.lastActiveDate = today;
      
      // Award daily bonus
      this.addXC(this.streak.nextBonus, 'Daily Login Bonus', 'daily_bonus');
      
      // Check streak achievement
      this.updateAchievement('daily_warrior', this.streak.currentStreak);
      
      this.saveState();
    }
  }

  // PUBLIC API
  getBalance(): number {
    return this.balance;
  }

  getTransactions(): XCTransaction[] {
    return this.transactions.sort((a, b) => b.timestamp - a.timestamp);
  }

  getAchievements(): XCAchievement[] {
    return this.achievements;
  }

  getStreak(): XCStreak {
    return this.streak;
  }

  addXC(amount: number, description: string, source: string): boolean {
    if (amount <= 0) return false;
    
    this.balance += amount;
    
    const transaction: XCTransaction = {
      id: Date.now().toString(),
      type: 'earn',
      amount,
      description,
      timestamp: Date.now(),
      source
    };
    
    this.transactions.push(transaction);
    
    // Update XC collector achievement
    this.updateAchievement('xc_collector', this.balance);
    
    this.saveState();
    this.notifyListeners('balanceUpdated', this.balance);
    this.notifyListeners('transactionAdded', transaction);
    
    return true;
  }

  spendXC(amount: number, description: string, source: string): boolean {
    if (amount <= 0 || this.balance < amount) return false;
    
    this.balance -= amount;
    
    const transaction: XCTransaction = {
      id: Date.now().toString(),
      type: 'spend',
      amount: -amount,
      description,
      timestamp: Date.now(),
      source
    };
    
    this.transactions.push(transaction);
    
    this.saveState();
    this.notifyListeners('balanceUpdated', this.balance);
    this.notifyListeners('transactionAdded', transaction);
    
    return true;
  }

  // EARNING METHODS
  awardDailyLogin(): void {
    this.checkDailyStreak();
  }

  awardBuzzPost(): void {
    this.addXC(5, 'Posted buzz message', 'buzz_post');
    this.updateAchievement('first_post', 1);
    this.updateAchievement('buzz_master', 1);
  }

  awardComment(): void {
    this.addXC(2, 'Posted comment', 'comment');
    this.updateAchievement('social_butterfly', 1);
  }

  awardInteraction(): void {
    this.addXC(1, 'Social interaction', 'interaction');
    this.updateAchievement('social_butterfly', 1);
  }

  awardMeshConnection(): void {
    this.addXC(15, 'Connected to mesh node', 'mesh_connection');
    this.updateAchievement('mesh_pioneer', 1);
  }

  awardChannelJoin(): void {
    this.addXC(3, 'Joined geohash channel', 'channel_join');
    this.updateAchievement('channel_hopper', 1);
  }

  awardProfileComplete(): void {
    this.addXC(20, 'Completed profile setup', 'profile_complete');
    this.updateAchievement('profile_complete', 1);
  }

  awardAchievement(achievementId: string): void {
    const achievement = this.achievements.find(a => a.id === achievementId);
    if (achievement && !achievement.completed) {
      this.addXC(achievement.reward, `Achievement: ${achievement.name}`, 'achievement');
      achievement.completed = true;
      this.saveState();
      this.notifyListeners('achievementCompleted', achievement);
    }
  }

  private updateAchievement(achievementId: string, increment: number): void {
    const achievement = this.achievements.find(a => a.id === achievementId);
    if (achievement && !achievement.completed) {
      achievement.progress = Math.min(achievement.progress + increment, achievement.maxProgress);
      
      if (achievement.progress >= achievement.maxProgress) {
        this.awardAchievement(achievementId);
      }
      
      this.saveState();
      this.notifyListeners('achievementUpdated', achievement);
    }
  }

  // EVENT LISTENERS
  subscribe(event: string, callback: (data: any) => void): () => void {
    if (!this.listeners[event]) {
      this.listeners[event] = [];
    }
    this.listeners[event].push(callback);
    
    return () => {
      this.listeners[event] = this.listeners[event].filter(cb => cb !== callback);
    };
  }

  private notifyListeners(event: string, data: any) {
    if (this.listeners[event]) {
      this.listeners[event].forEach(callback => callback(data));
    }
  }

  // ADMIN METHODS (for testing/debugging)
  debugAddXC(amount: number): void {
    this.addXC(amount, 'Debug addition', 'debug');
  }

  getStats() {
    return {
      balance: this.balance,
      totalEarned: this.transactions.filter(t => t.type === 'earn').reduce((sum, t) => sum + t.amount, 0),
      totalSpent: Math.abs(this.transactions.filter(t => t.type === 'spend').reduce((sum, t) => sum + t.amount, 0)),
      transactionCount: this.transactions.length,
      achievementsCompleted: this.achievements.filter(a => a.completed).length,
      currentStreak: this.streak.currentStreak
    };
  }
}

export const xcEconomy = new XCEconomyService();
